provider/ansible.py

```diff
@@ -167,16 +172,28 @@ def distro_install(packages="ansible"):
     policy_map = {"distro_install": distro_install,
                   "python_install": python_install}
 
+    ansible_install_policy = params['ansible_install_policy']
+    if ansible_install_policy not in policy_map:
+        LOG_JOB.error(f"No valid install policy: {ansible_install_policy}.")
+        return False
```

Comment:
Here I want to add a comment. When `ansible_install_policy` is not defined but `ansible-playbook` is available, the method needs to return True. So if `check_cmd`, go to next step, if no, `path.find_command('ansible-playbook')`.

Reply:
@PaulYuuu I think I follow, just let me confirm one thing: Is it because of the following test cases?
```
nested_block_resize.py
ansible_with_responsive_migration.py
```
Configuration files for these two doesn't define `ansible_install_policy` at all – compared to `ansible_test`, which is the only test containing this configuration parameter.

Am I correct?

Reply:
These changes should resolve it:
```
    ansible_install_policy = params.get('ansible_install_policy')
    if ansible_install_policy:
        if ansible_install_policy not in policy_map:
            LOG_JOB.error(f"No valid install policy: {ansible_install_policy}.")
            return False
    package_list = params.get_list("package_list", 'sshpass')
    try:
        if ansible_install_policy == 'python_install':
            pip_bin = _pip_binary()
            if not pip_bin:
                return False
            check_cmd = rf"{pip_bin} freeze | grep -v ansible-core | grep -q ansible="
        elif ansible_install_policy == 'distro_install':
            check_cmd = params.get("ansible_check_cmd")
            package_list.insert(0, 'ansible')
        if "check_cmd" in locals():
            LOG_JOB.debug(f"Is full ansible version installed: '{check_cmd}'")
            process.run(check_cmd, verbose=False, shell=True)
        else:
            path.find_command('ansible-playbook')
```
1. `params[]` is replaced by `params.get()` because parameter can be undefined
2.  Following syntax bellow will ensure we return False only if `ansible_install_policy` contains some nonsense. If it is not defined at all nothing happens.
 ```
    if ansible_install_policy:
        if ansible_install_policy not in policy_map:
            LOG_JOB.error(f"No valid install policy: {ansible_install_policy}.")
            return False
```
4. `if check_cmd` is replaced by `if "check_cmd" in locals()` because `check_cmd` can be undefined

Reply:
(I did a quick test with no `ansible_install_policy` defined and with previously installed ansible on the system, so it relies only on `ansible-playbook` check and it works as expected.)

Reply:
Not exactly. Users only need to install ansible once, so if the system has ansible, or runs ansible_test in a loop, we don't need to go to the install policy in each case.

On the other hand, to make the code simple, we can define the global `check_cmd` parameter instead of using `if "check_cmd" in locals()`.
```python
try:
    ansible_install_policy = params.get('ansible_install_policy')
    check_cmd = params.get('ansible_check_cmd')
    if check_cmd and ansible_install_policy in policy_map:
        if process.system(check_cmd, verbose=False, shell=True):
            return policy_map[ansible_install_policy]()
    else:
        path.find_command('ansible-playbook')
except:
    return False
return True
```

This is the basic code, you can modify it to map check_cmd with different policies.

Reply:
I really like the idea of simplification because the logic of ansible provider is starting to be a bit more complex than intended I think :slightly_smiling_face: But regarding install policy, it was already achieved by the previous code: if no `ansible_install_policy` is defined we stick to the `ansible-playbook` and no installation of ansible is enforced. 

Also there is a problem with `check_cmd`: `distro_install` policy must run properly if the `check_cmd` is defined and if it is not, and it is also complicates the code.

I prepared two versions with modifications. Both does the same:

**Version A:**

```python
    ansible_install_policy = params.get('ansible_install_policy')
    if ansible_install_policy:
        if ansible_install_policy not in policy_map:
            LOG_JOB.error(f"No valid install policy: {ansible_install_policy}.")
            return False
    package_list = params.get_list("package_list", 'sshpass')
    try:
        check_cmd = params.get("ansible_check_cmd")
        if ansible_install_policy == 'python_install':
            pip_bin = _pip_binary()
            if not pip_bin:
                return False
            check_cmd = rf"{pip_bin} freeze | grep -v ansible-core | grep -q ansible="
        elif ansible_install_policy == 'distro_install':
            package_list.insert(0, 'ansible')
        if check_cmd in locals():
            LOG_JOB.debug(f"Is full ansible version installed: '{check_cmd}'")
            process.run(check_cmd, verbose=False, shell=True)
        else:
            path.find_command('ansible-playbook')
```

**Version B:**

```python
    package_list = params.get_list("package_list", 'sshpass')
    try:
        ansible_install_policy = params.get('ansible_install_policy')
        if ansible_install_policy:
            check_cmd = params.get("ansible_check_cmd")
            if ansible_install_policy == 'python_install':
                pip_bin = _pip_binary()
                if not pip_bin:
                    return False
                check_cmd = rf"{pip_bin} freeze | grep -v ansible-core | grep -q ansible="
            elif ansible_install_policy == 'distro_install':
                package_list.insert(0, 'ansible')
            else:
                LOG_JOB.error(f"No valid install policy: {ansible_install_policy}.")
                return False
            if check_cmd:
                LOG_JOB.debug(f"Is full ansible version installed: '{check_cmd}'")
                process.run(check_cmd, verbose=False, shell=True)
        else:
            path.find_command('ansible-playbook')
```
What do you think about these? I personally prefer the A version which is just a slight modification of the previous code and IMHO it is easier to read.

Reply:
@yanan-fu @PaulYuuu Thank you both for discussion! I incorporated the remaining changes (regarding except block) into the variant A and I am going to test these changes ASAP.